### PR TITLE
Enhance repo_type_and_id_from_hf_id of hf_api

### PR DESCRIPTION
**Pull request by pulltheflower**
Related to issue https://github.com/huggingface/huggingface_hub/issues/3494

### CODE REVIEW

**Comment by Wauplin** [comment]
@bot /style

**Comment by pulltheflower** [comment]
@Wauplin I noticed a large number of build and test failures, which seem to be related to network issues. Could this possibly be connected to my changes? I’d appreciate it if you could take a look, thank you!

**Comment by Wauplin** [comment]
> I noticed a large number of build and test failures, which seem to be related to network issues. 

Yes that was due to temporary issues on our staging infra. It is now resolved.

Note that there are still a few issues due to this PR. See all the 

```
FAILED ../tests/test_hf_api.py::RepoUrlTest::test_repo_url_class - ValueError: not enough values to unpack (expected 2, got 1)
FAILED ../tests/test_hf_api.py::ParseHFUrlTest::test_repo_type_and_id_from_hf_id_on_wrong_values - AssertionError: ValueError not raised
FAILED ../tests/test_hf_api.py::RepoUrlTest::test_repo_url_endpoint - ValueError: not enough values to unpack (expected 2, got 1)
FAILED ../tests/test_hf_api.py::RepoUrlTest::test_repo_url_namespace - ValueError: not enough values to unpack (expected 2, got 1)
```

looking into the details, [this line](https://github.com/huggingface/huggingface_hub/pull/3507/files#diff-71a3e23532e1355e79edd51b351bd514f0aa55b946e4c530b0c95f647773ac5dL251) is now failing.

Could you have a look at it please?

File: tests/test_hf_api.py [hunk 1]
```diff
@@ -2760,6 +2760,9 @@ def test_git_push_end_to_end(self):
 class ParseHFUrlTest(unittest.TestCase):
     def test_repo_type_and_id_from_hf_id_on_correct_values(self):
         possible_values = {
+            "http://localhost:8080/hf/user/id": [None, "user", "id"],
+            "http://localhost:8080/hf/datasets/user/id": ["dataset", "user", "id"],
+            "http://localhost:8080/hf/models/user/id": ["model", "user", "id"],
```
Comment (julien-c) [inline tests/test_hf_api.py]: move those at the end given it's an edge case

Comment (pulltheflower) [reply tests/test_hf_api.py]: Sorry for the late response. I’ve made the modification.

**Comment by pulltheflower** [comment]
@Wauplin Sorry for taking so long to fix the tests. Could you please trigger the pipeline again?

**Comment by pulltheflower** [comment]
@Wauplin There were still some failing tests in the previous commit, but I’ve fixed them in the latest commit.

**Comment by pulltheflower** [comment]
image

@Wauplin The test is passing successfully in my local environment but failing in the CI/CD pipeline. Could you please take a look and help identify what might be causing this discrepancy?

File: tests/test_hf_api.py [hunk 2]
```diff
@@ -2758,25 +2758,34 @@ def test_git_push_end_to_end(self):
...
-            "https://huggingface.co/spaces/user/id": ["space", "user", "id"],
-            "user/id": [None, "user", "id"],
-            "dataset/user/id": ["dataset", "user", "id"],
-            "space/user/id": ["space", "user", "id"],
-            "id": [None, None, "id"],
-            "hf://id": [None, None, "id"],
-            "hf://user/id": [None, "user", "id"],
-            "hf://model/user/name": ["model", "user", "name"],  # 's' is optional
-            "hf://models/user/name": ["model", "user", "name"],
+            "saas": {
```
Comment (Wauplin) [inline tests/test_hf_api.py]:
```suggestion
            "hub": {
```

File: tests/test_hf_api.py [hunk 3]
```diff
@@ -2758,25 +2758,34 @@ def test_git_push_end_to_end(self):
...
+                "http://localhost:8080/hf/models/user/id": ["model", "user", "id"],
+            },
         }
 
         for key, value in possible_values.items():
-            self.assertEqual(
-                repo_type_and_id_from_hf_id(key, hub_url=ENDPOINT_PRODUCTION),
-                tuple(value),
-            )
+            hub_url = ENDPOINT_PRODUCTION if key == "saas" else "http://localhost:8080/hf"
```
Comment (Wauplin) [inline tests/test_hf_api.py]:
```suggestion
            hub_url = ENDPOINT_PRODUCTION if key == "hub" else "http://localhost:8080/hf"
```

File: tests/test_hf_api.py [hunk 4]
```diff
@@ -2758,25 +2758,34 @@ def test_git_push_end_to_end(self):
...
+            for key, value in value.items():
+                self.assertEqual(
+                    repo_type_and_id_from_hf_id(key, hub_url=hub_url),
+                    tuple(value),
+                )
```
Comment (Wauplin) [inline tests/test_hf_api.py]:
```suggestion
                assert repo_type_and_id_from_hf_id(key, hub_url=hub_url) == tuple(value)
```

(not related to this PR but we are slowly removing the `self.assertEqual` from the codebase)

File: src/huggingface_hub/hf_api.py [hunk 5]
```diff
@@ -238,28 +238,62 @@ def repo_type_and_id_from_hf_id(hf_id: str, hub_url: Optional[str] = None) -> tu
     """
     input_hf_id = hf_id
 
-    hub_url = re.sub(r"https?://", "", hub_url if hub_url is not None else constants.ENDPOINT)
-    is_hf_url = hub_url in hf_id and "@" not in hf_id
+    # Get the hub_url (with or without protocol)
+    full_hub_url = hub_url if hub_url is not None else constants.ENDPOINT
+    hub_url_without_protocol = re.sub(r"https?://", "", full_hub_url)
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py]: Slight optimization but could you define a root-level `_REGEX_HTTP_PROTOCOL = re.compile(r"https?://")` variable (next to `_REGEX_DISCUSSION_URL` for instance) and then use

```
hub_url_without_protocol = _REGEX_HTTP_PROTOCOL.sub("", full_hub_url)
```

It prevents Python from having to compile the regex on each call

File: src/huggingface_hub/hf_api.py [hunk 6]
```diff
@@ -238,28 +238,62 @@ def repo_type_and_id_from_hf_id(hf_id: str, hub_url: Optional[str] = None) -> tu
...
+
+    # Check if hf_id is a URL containing the hub_url (check both with and without protocol)
+    hf_id_without_protocol = re.sub(r"https?://", "", hf_id)
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py]: reused here

File: src/huggingface_hub/hf_api.py [hunk 7]
```diff
@@ -238,28 +238,62 @@ def repo_type_and_id_from_hf_id(hf_id: str, hub_url: Optional[str] = None) -> tu
...
+    is_hf_url = hub_url_without_protocol in hf_id_without_protocol and "@" not in hf_id
 
     HFFS_PREFIX = "hf://"
     if hf_id.startswith(HFFS_PREFIX):  # Remove "hf://" prefix if exists
         hf_id = hf_id[len(HFFS_PREFIX) :]
 
+    # If it's a URL, strip the endpoint prefix to get the path
+    if is_hf_url:
+        # Remove protocol if present
+        hf_id_normalized = re.sub(r"https?://", "", hf_id)
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py]: and here

File: src/huggingface_hub/hf_api.py (hunk 5 above)
Comment (pulltheflower) [reply src/huggingface_hub/hf_api.py]: Thanks for the review! I've made the requested changes. Please take another look when you have time.
